### ISSUE TITLE
Flake falsely assumes the store path.

### ISSUE THREAD

**Issue by LEXUGE** [NixOS/nix#4081]
**Describe the bug**

Recently, I have been trying to install my full configuration in flake at once from NixOS LiveCD. To achieve that, I have to use the store on the target drive instead of that on tmpfs or OverlayFS.

However, even if I apply `--option store /mnt`, `nix` sometimes still uses `/nix/store` instead of `/mnt/nix/store` to check and eval.

**Steps To Reproduce**

1. `nix build "${MOUNTPOINT}/etc/nixos#name" --option store ${MOUNTPOINT}`

Above would throw error like
```
path /nix/store/nvmqjsr9nh3vy7w2p6bqqmrjs17q5dfc-nix-3.0pre20200829_f156513/ is not valid
```

To get around it, I have to `nix copy --to /mnt nixpkgs#nixFlakes --no-check-sigs` (without `--no-check-sigs` it doesn't work on LiveCD, despite normal behavior on an ordinary NixOS).

2. After `nix copy`, run above `nix build` again gives me complaints about absence of `flake.nix` (I guess this is because `nix` archived the `/mnt/etc/nixos` to `/mnt/nix/store`, *but* still checked presence of `flake.nix` in `/nix/store`).

3. If I manually `nix archive` it, and run `nix build` again, it will throw error about impurity.
```
access to path `/mnt/nix/store/some_hash_here-source/flake.nix` is forbidden in restricted mode.
```
which I guess is caused by `nix` thinking `/mnt/nix/store` is not the store (anything other than `/nix/store` is not allowed then).

**Expected behavior**

Use `/mnt/nix/store` for all occasions.

**`nix-env --version` output**
nix-env (Nix) 3.0pre20200829_f156513
**Additional context**

**Comment by xaverdh** [comment]
As a workaround you can use the `--impure` flag.
I think a simpler way to reproduce the core problem is trying to run something like

```sh
nix build --option store ./test-store nixpkgs#hello
```

which will fail with an `InvalidPath` error (due to nix itself).

**Comment by xaverdh** [comment]
on a related note, the out link for local stores also appears to be wrong, likely for the same reason..

**Comment by LEXUGE** [comment]
Yes, if you try to use the external store, the result link is still `/nix/store`.

**Comment by xaverdh** [comment]
Suggestion: maybe rename this issue to something like "local stores don't work in pure evaluation mode" or "invalid path error with local store"?
At least to me it wasn't quite obvious from the title what this is about..

**Comment by LEXUGE** [comment]
Well, but I do think this issue is related to other references to the store in the codebase because nix is getting wrong store paths in several occasions as I described above.

**Comment by xaverdh** [comment]
Not sure if this issue is getting as much attention as it should.. @edolstra are you aware of this?
(sorry if I am too pushy)

**Comment by Ericson2314** [comment]
I think the issue is the evaluation cache writing down paths. One thing I did in #3754 is is make it write down "store path descriptors" (hashes and references used to write the store path) instead of the store path itself. This means a store path can be computed from that from any store dir.

**Comment by shadowrylander** [comment]
Any updates on this? I got past the original "fetched file isn't in nix store" by using the latest master and setting the store to `/` (i.e. `/nix/store`), but now I'm stuck on `chroot: failed to run command '/nix/var/nix/profiles/system/activate': No such file or directory` and `chroot: failed to run command '/run/current-system/bin/switch-to-configuration': No such file or directory`; I'm assuming this is related, as the `/mnt` store hasn't been modified, so chrooting into `/mnt` wouldn't find the appropriate files?

**Comment by xaverdh** [comment]
If I am not mistaken https://github.com/NixOS/nix/commit/4badb6943f73e3d740d4de793eb9860e26842bc1 fixes part of this. Unfortunately even with this commit the symlink will point to the correct path, so uses of these symlinks in nixos-rebuild / install might still cause trouble.

**Comment by shadowrylander** [comment]
I can rebuild my spare system with the latest master and report back on whether that still occurs or not, if you wish?

**Comment by shadowrylander** [comment]
@xaverdh Well that didn't work, though I'm not sure I did it properly to begin with; how do I use the master branch of the Nix repo in my flake?

**Comment by shadowrylander** [comment]
Never mind; I figured out how to use the master branch, but now I'm getting stuck on `'flake.nix' file of flake [flake] escapes from [path]`.

**Comment by shadowrylander** [comment]
@xaverdh So I can use absolute paths everywhere?

**Comment by shadowrylander** [comment]
Any updates on this, if I may ask?

**Comment by shadowrylander** [comment]
May I ask if there are any updates on this again? I am currently attempting to initialize a new system with the flake, but am getting the same errors as above. Should I just create a smaller configuration without flakes to initialize the new system in the meantime?

**Comment by cole-h** [comment]
FWIW, I also ran into the same issue. Thanks to a pointer on IRC from @colemickens, I was able to get it up and running.

`nix build .#nixosConfigurations.hostname && nixos-install --system ./result` is essentially what I had to do.

**Comment by shadowrylander** [comment]
@cole-h Which is similar to the advice given [here](https://github.com/hlissner/dotfiles/issues/11#issuecomment-739152960); however, I'm still getting errors where any repos referenced using flake paths are not being found, such as when using `{ inputs.<input>.flake = false; }` in a flake.

**Comment by figsoda** [comment]
still important to me

**Comment by domenkozar** [comment]
Detailed logs are in https://github.com/NixOS/nixpkgs/issues/126141#issuecomment-879666206

**Comment by bhankas** [comment]
I had to use solution from @cole-h 's comment today while setting up my Raspberry Pi. The error message is not very helpful, and the solution is not obvious. Fix does become semi-obvious in hindsight, but it is a UX gap, and might be worth fixing.
